### PR TITLE
SA1127GenericTypeConstraintsMustBeOnOwnLine

### LINKED ISSUE #1176: New rule: Generic type constraints must be on their own line
**Category:** Readability
**Title:** Generic type constraints must be on own line
**Description:** For maximum readability, the constraints clause for a generic type parameter should be listed on a line of code by themselves. The following code would produce this violation:

```csharp
private void Method<T>() where T : class
{
}
```

To fix a violation of this rule, place the constraints for each generic type on their own line:

```csharp
private void Method<T>()
    where T : class
{
}
```

:memo: For types with more than one generic type parameter, each parameter would need to appear on a line by itself. For example, the following shows the correct way to write three constraints:

```csharp
private void Method<T1, T2, T3>()
    where T1 : class
    where T2 : class
    where T3 : class
{
}
```

### PR DESCRIPTION
**Pull request by nbarbettini**
Fixes #1176 

Didn't realize I was supposed to open up a separate issue for discussing the code fix. Oops, will do in the future.

The diagnostic part was pretty easy. Any feedback is helpful re: how I went about the code fix. If I'm doing something really weird or non-idiomatic to Roslyn analyzers, I can go back and fix it.

I tried to cover as many scenarios as I could think of in the tests; hopefully it shows that the code fix is relatively stable.

### CODE REVIEW

**Comment by sharwell** [comment]
No need to create a separate issue if you already have a pull request ready. :smile:

**Comment by nbarbettini** [comment]
Cool @sharwell. I just noticed the CI build failing... I didn't see the change to `RegisterSyntaxNodeActionHonorExclusions` before I created the PR. Just a min...

File: documentation/SA1127.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+## How to fix violations
+
+To fix a violation of this rule, move each `where` clause to its own line.
+
+## How to suppress violations
+
+```csharp
+[SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1127:GenericTypeConstraintsMustBeOnOwnLine", Justification = "Reviewed.")]
```
Comment (sharwell) [inline documentation/SA1127.md]: :bulb: You can omit this suppression method.

File: documentation/SA1127.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+```csharp
+#pragma warning disable SA1127 // GenericTypeConstraintsMustBeOnOwnLine
```
Comment (sharwell) [inline documentation/SA1127.md]: :bulb: Can you add example code within the suppressed section which would normally violate this rule?

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1127GenericTypeConstraintsMustBeOnOwnLine.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            }
+        }
+
+        /// <inheritdoc/>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(this.HandleCompilationStart);
+        }
+
+        private void HandleCompilationStart(CompilationStartAnalysisContext context)
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1127GenericTypeConstraintsMustBeOnOwnLine.cs]: :bulb: This method can be static.

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1127GenericTypeConstraintsMustBeOnOwnLine.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+{
+    using System.Collections.Immutable;
+    using Helpers;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    /// <summary>
+    /// Todo
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1127GenericTypeConstraintsMustBeOnOwnLine.cs]: :exclamation: Todo.

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1127CodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                if (trivia.IsKind(SyntaxKind.WhitespaceTrivia))
+                {
+                    parentIndentation += trivia.ToString();
+                }
+            }
+
+            return parentIndentation;
+        }
+
+        // This function will remove any unnecessary whitespace or end-of-line trivia from a token.
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1127CodeFixProvider.cs]: :bulb: This is better as a proper documentation comment.

**Comment by nbarbettini** [comment]
@sharwell I believe I have addressed all of your comments :+1:

File: documentation/SA1127.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+{
+}
+```
+
+The following code would not produce any violations:
+
+```csharp
+private void Method<T, R>()
+    where T : class
+	where R : class, new()
```
Comment (sharwell) [inline documentation/SA1127.md]: :bulb: this line can be indented

File: documentation/SA1127.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+## How to suppress violations
+
+```csharp
+#pragma warning disable SA1127 // Generic type constraints must be on their own line
+private void Method<T>() where T : class
+{
+	// method body...
+}
+#pragma warning restore SA1127 // Generic type constraints must be on their own line
```
Comment (sharwell) [inline documentation/SA1127.md]: :bulb: This restore line can be moved up above the `{`.

**Comment by sharwell** [comment]
:question: This looks good overall, but I'd like to rebase/squash it before merging it. Are you comfortable doing this operation, or would you like for me to just do it when I merge the code?

**Comment by nbarbettini** [comment]
I'm a TFS guy who recently moved to Github, so I'm a newbie. :) Can you rebase this one, and I'll be sure to do this locally before doing PRs in the future?

**Comment by sharwell** [comment]
@nbarbettini Rebasing/squashing generally isn't required, but in this case there's a bit of back-and-forth in the commit history that I was looking to consolidate. :smile:

**Comment by nbarbettini** [comment]
Yeah :grimacing: Turns out the git shell is a little more reliable than the VS git plugin. Bear with me, I'll get the hang of this.

**Comment by sharwell** [comment]
@nbarbettini The tools I use are:

* VS Git support (the one that comes with VS): For seeing file status in Solution Explorer (that's it)
* Git Diff Margin: For seeing file status in the line numbers margin and overlay on the scroll bar
* GitExtensions: For almost everything else (staging, fetching, pushing, merging, rebasing, etc...)
* BeyondCompare 4: Diff/merge

I don't like wasting time dealing with tools, and these seem to keep up with my workflow pretty well.

**Comment by sharwell** [comment]
Rebased and merged as commit d3ed286dc5465ffee2a67d9c4c2ac3018af15a04.

**Pull request closed without merging**